src/python/pants/backend/python/subsystems/pex_build_util.py

```diff
@@ -318,13 +347,99 @@ def _prepare_inits(self) -> Set[str]:
+
+    See this test case as background:
+    https://github.com/pantsbuild/pex/commit/66aa5ef69e8b65004a548b44edc0b81dfa8ec736#diff-b5b57c9d4fc663f59ecaedb851c13bd8R23
+    """
+    ipex_info = self._builder.info.copy()
+    orig_chroot = self._builder.chroot()
+    # Replace the original builder with a new one, and just pull files from the old chroot.
+    self._builder = PEXBuilder(interpreter=self._builder.interpreter)
+
+    # If the dependencies are very large, this flag is indicated to be useful in RAM-constrained
+    # environments.
+    self._builder.info.always_write_cache = True
+
+    # PTEX-INFO: Contains resolver settings and source files to copy over into the dehydrated .ipex
+    #            file.
+    all_code = []
+    for label in ['source', 'resource']:
+      for src in orig_chroot.get(label):
+        if src in ['__main__.py', PexInfo.PATH]:
+          continue
+        all_code.append(src)
+        self._builder.chroot().copy(os.path.join(orig_chroot.path(), src), src, label=label)
+
+    python_setup = self._python_setup_subsystem
+    python_repos = self._python_repos_subsystem
+
+    resolver_settings = dict(
+      indexes=python_repos.indexes,
+      find_links=list(self._all_find_links),
+      allow_prereleases=bool(python_setup.resolver_allow_prereleases),
+    )
+
+    ptex_info = dict(
+      code=all_code,
+      resolver_settings=resolver_settings,
+      transitive_requirements_by_platform={
+        (
+          platform if platform != 'current' else Platform.current().platform
+        ): reqs
+        for platform, reqs in self._all_transitive_requirements_by_platform.items()
+      },
+    )
+    with temporary_file(permissions=0o644) as ptex_info_file:
+      ptex_info_file.write(json.dumps(ptex_info).encode())
+      ptex_info_file.flush()
+      self._builder.add_resource(filename=ptex_info_file.name, env_filename='PTEX-INFO')
+
+    # IPEX-INFO: The original PEX-INFO, which should be the PEX-INFO in the hydrated .pex file that
+    #            is generated when the .ipex is first executed.
+    with temporary_file(permissions=0o644) as ipex_info_file:
+      ipex_info_file.write(ipex_info.dump().encode())
+      ipex_info_file.flush()
+      self._builder.add_resource(filename=ipex_info_file.name, env_filename='IPEX-INFO')
+
+    # _ptex_launcher.py: The special bootstrap script to hydrate the .ipex with the fully resolved
+    #                    requirements when it is first executed.
+    parent_module = 'pants.backend.python.subsystems.pex_bootstrap'
+    ptex_launcher = get_provider(parent_module).get_resource_string(parent_module, '_ptex_launcher.py')
+    with temporary_file(permissions=0o644) as ptex_launcher_file:
+      ptex_launcher_file.write(ptex_launcher)
+      ptex_launcher_file.flush()
+      self._builder.add_source(filename=ptex_launcher_file.name, env_filename='_ptex_launcher.py')
+    # Our .ptex file will use our special preamble.
```

Comment:
Given a preamble is actually a thing in pex, this comment seems misleading? Maybe just remove the comment?

Reply:
Have reworded this comment to describe it as a "custom app launcher"!